google/cloud/spanner/internal/connection_impl.h

```diff
@@ -0,0 +1,77 @@
+// Copyright 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef GOOGLE_CLOUD_CPP_SPANNER_GOOGLE_CLOUD_SPANNER_INTERNAL_CONNECTION_IMPL_H_
+#define GOOGLE_CLOUD_CPP_SPANNER_GOOGLE_CLOUD_SPANNER_INTERNAL_CONNECTION_IMPL_H_
+
+#include "google/cloud/spanner/connection.h"
+#include "google/cloud/spanner/internal/spanner_stub.h"
+#include "google/cloud/spanner/version.h"
+#include "google/cloud/status_or.h"
+#include <memory>
+#include <string>
+
+namespace google {
+namespace cloud {
+namespace spanner {
+inline namespace SPANNER_CLIENT_NS {
+namespace internal {
+
+// A concrete `Connection` subclass that uses gRPC to actually talk to a real
+// Spanner instance. See `MakeConnection()` for a factory function that creates
+// and returns instances of this class.
+class ConnectionImpl : public Connection {
+ public:
+  // Creates a ConnectionImpl that will talk to the specified `database` using
+  // the given `stub`. We can test this class by injecting in a mock `stub`.
+  explicit ConnectionImpl(std::string database,
+                          std::shared_ptr<internal::SpannerStub> stub)
+      : database_(std::move(database)), stub_(std::move(stub)) {}
+
+  StatusOr<CommitResult> Commit(CommitParams) override;
+
+ private:
+  class SessionHolder {
+   public:
+    SessionHolder(std::string session, ConnectionImpl* conn) noexcept
+        : session_(std::move(session)), conn_(conn) {}
+    ~SessionHolder() { conn_->sessions_.emplace_back(std::move(session_)); }
+
+    std::string const& session_name() const { return session_; }
+
+   private:
+    std::string session_;
+    ConnectionImpl* conn_;
+  };
+  friend class SessionHolder;
+  StatusOr<SessionHolder> GetSession();
+
+  /// Implementation details for Commit.
+  StatusOr<CommitResult> Commit(google::spanner::v1::TransactionSelector& s,
+                                std::vector<Mutation> mutations);
+  std::string database_;
+  std::shared_ptr<internal::SpannerStub> stub_;
```

Comment:
Why do we choose to make this shared?  gRPC gave it to us as a `unique_ptr`, which we converted into a `shared_ptr`, but (as far as I can see) its refcount remains at 1.

Reply:
> Don't know. I'm just copying existing code. Do you mind if we follow up on this change in a separate PR?

I'd suggest that it should be unique until it needs to be otherwise, but a separate PR is good.